File: packages/sdk/src/resources/artifacts.ts
import type { ArtifactId } from "@osva/contracts";
import {
  artifactListResourceSchema,
  artifactResourceSchema,
  listArtifactsQuerySchema,
} from "@osva/contracts/schemas";
import type { z } from "zod";

import type { DownloadResult, OsvaHttpClient } from "../http-client.js";

type ArtifactResource = z.infer<typeof artifactResourceSchema>;
type ArtifactListResource = z.infer<typeof artifactListResourceSchema>;
type ListArtifactsQuery = z.infer<typeof listArtifactsQuerySchema>;

export interface CreateArtifactInput {
  readonly name: string;
  readonly content: Blob;
  readonly mediaType?: string;
  readonly metadata?: Record<string, unknown>;
  readonly expectedDigest?: string;
  readonly idempotencyKey?: string;
}

export class ArtifactsResource {
  constructor(private readonly client: OsvaHttpClient) {}

  list(query: ListArtifactsQuery = {}): Promise<ArtifactListResource> {
    return this.client.request({
      method: "GET",
      path: "/v1/artifacts",
      query: {
        runId: query.runId,
        runAttemptId: query.runAttemptId,
        limit: query.limit,
        cursor: query.cursor,
      },
    });
  }

  get(artifactId: ArtifactId): Promise<ArtifactResource> {
    return this.client.request({
      method: "GET",
      path: `/v1/artifacts/${encodeURIComponent(artifactId)}`,
    });
  }

  create(input: CreateArtifactInput): Promise<ArtifactResource> {
    const form = new FormData();
    form.set("name", input.name);
    if (input.mediaType !== undefined) {
      form.set("mediaType", input.mediaType);
    }
    if (input.metadata !== undefined) {
      form.set("metadata", JSON.stringify(input.metadata));
    }
    if (input.expectedDigest !== undefined) {
      form.set("expectedDigest", input.expectedDigest);
    }
    if (input.idempotencyKey !== undefined) {
      form.set("idempotencyKey", input.idempotencyKey);
    }
    form.set("file", input.content, input.name);

    const headers: Record<string, string> = {};
    if (input.idempotencyKey !== undefined) {
      headers["Idempotency-Key"] = input.idempotencyKey;
    }

    return this.client.uploadMultipart({
      path: "/v1/artifacts",
      form,
      headers,
    });
  }

  download(artifactId: ArtifactId): Promise<DownloadResult> {
    return this.client.download({
      path: `/v1/artifacts/${encodeURIComponent(artifactId)}/content`,
    });
  }
}

File: sdks/python/src/osva/resources/artifacts.py
from __future__ import annotations

from typing import Any

import httpx

from osva.http_client import HTTPClient


class ArtifactsResource:
    def __init__(self, http: HTTPClient) -> None:
        self._http = http

    def list(self, **query: str | None) -> dict[str, Any]:
        return self._http.request("GET", "/v1/artifacts", query=query)

    def get(self, artifact_id: str) -> dict[str, Any]:
        return self._http.request("GET", f"/v1/artifacts/{artifact_id}")

    def create(
        self,
        *,
        name: str,
        content: bytes,
        media_type: str | None = None,
        metadata: dict[str, Any] | None = None,
        expected_digest: str | None = None,
        idempotency_key: str | None = None,
    ) -> dict[str, Any]:
        data: dict[str, str] = {"name": name}
        if media_type is not None:
            data["mediaType"] = media_type
        if metadata is not None:
            data["metadata"] = __import__("json").dumps(metadata)
        if expected_digest is not None:
            data["expectedDigest"] = expected_digest
        if idempotency_key is not None:
            data["idempotencyKey"] = idempotency_key

        files: dict[str, tuple[str, bytes, str | None]] = {
            "file": (
                name,
                content,
                media_type or "application/octet-stream",
            )
        }
        headers = {"Idempotency-Key": idempotency_key} if idempotency_key is not None else None
        return self._http.upload_multipart(
            "/v1/artifacts",
            data=data,
            files=files,
            headers=headers,
        )

    def download(self, artifact_id: str) -> httpx.Response:
        return self._http.download(f"/v1/artifacts/{artifact_id}/content")
